### ISSUE TITLE
[question] opencv with VideoCapture 

### ISSUE THREAD

**Issue by muhflare** [conan-io/conan#11541]
I am new to conan package manager, through some cpp files that I have written I tried to use opencv library and specifically the VideoCapture for reading videos, the thing is I did not succeed in any way to make the file compile and more specifically in linking my main.cpp with the required libs.
my conanfile.txt is :
[requires]
opencv/4.5.5
aws-sdk-cpp/1.9.234
openssl/1.1.1n

[generators]
compiler_args

and then I created a makefile like this:
all: conan.install compile
	
conan.install:
	conan install . --build=missing --install-folder build
	
compile:
	g++ src/main.cpp @build/conanbuildinfo.args -o bin/argss -std=c++17

the main.cpp contains only some lines:

#include <opencv2/core.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>

#include <opencv2/opencv.hpp>

#include <iostream>

using namespace std;
using namespace cv;

int main()
{
    VideoCapture cap;
    cout << "All good!!!" << endl;
    return 0;
}

but when I tried to run make all I got the following:

/usr/bin/ld: /home/muhammad/.conan/data/libx264/20191217/_/_/package/52bfaf5780e9ee7fba3fa4a492156dbcf0dff057/lib/libx264.a(analyse-8.o): in function `x264_8_analyse_init_costs':
analyse.c:(.text+0x216c9): undefined reference to `__log2f_finite'
/usr/bin/ld: /home/muhammad/.conan/data/libx264/20191217/_/_/package/52bfaf5780e9ee7fba3fa4a492156dbcf0dff057/lib/libx264.a(analyse-10.o): in function `x264_10_analyse_init_costs':
analyse.c:(.text+0x22dd9): undefined reference to `__log2f_finite'
/usr/bin/ld: /home/muhammad/.conan/data/libx265/3.4/_/_/package/e4cbe407e66aa4f34e55a1264d4af327f26b5f7d/lib/libx265.a(encoder.cpp.o): in function `x265::Encoder::ComputePSNR(x265_picture*, x265_picture*, x265_param*)':
encoder.cpp:(.text+0x2c2a): undefined reference to `__log10_finite'
/usr/bin/ld: encoder.cpp:(.text+0x2db0): undefined reference to `__log10_finite'
/usr/bin/ld: encoder.cpp:(.text+0x2e0e): undefined reference to `__log10_finite'
/usr/bin/ld: /home/muhammad/.conan/data/libx265/3.4/_/_/package/e4cbe407e66aa4f34e55a1264d4af327f26b5f7d/lib/libx265.a(encoder.cpp.o): in function `x265::Encoder::finishFrameStats(x265::Frame*, x265::FrameEncoder*, x265_frame_stats*, int)':
encoder.cpp:(.text+0x53e9): undefined reference to `__log10_finite'
/usr/bin/ld: encoder.cpp:(.text+0x5442): undefined reference to `__log10_finite'

and several lines with a similar content

am I doing something wrong in here????
what should I provide so that it can be compiled and used???
- [ ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @muhflare 

It would be great to see the full command and terminal output when you do ``conan install``. That output will contain for example some prints of your default profile used to install the binaries. One of the possibilities is that your current compiler configuration is not matching the one use to retrieve the packages (maybe the C++=17 standard? Not a typical source of binary incompatibility, but could be)

I guess the other deps, besides opencv can be removed from the above ``conanfile.txt`` and still will behave the same.

Can you also please report your OS, version, compiler, compiler version?

**Comment by muhflare** [comment]
[out.txt](https://github.com/conan-io/conan/files/9002176/out.txt)
this is the ourput of the `$ conan install . --build=missing --install-folder build`
as for `conanfile.txt` I am interested in keeping then as I want to use them later on

my OS:
`NAME="Linux Mint"
VERSION="20.3 (Una)"
ID=linuxmint
ID_LIKE=ubuntu
PRETTY_NAME="Linux Mint 20.3"
VERSION_ID="20.3"
HOME_URL="https://www.linuxmint.com/"
SUPPORT_URL="https://forums.linuxmint.com/"
BUG_REPORT_URL="http://linuxmint-troubleshooting-guide.readthedocs.io/en/latest/"
PRIVACY_POLICY_URL="https://www.linuxmint.com/"
VERSION_CODENAME=una
UBUNTU_CODENAME=focal
`
g++ version: 9.4.0

**Comment by memsharded** [comment]
> as for conanfile.txt I am interested in keeping then as I want to use them later on

Yes, sure, but for reporting issues, it is always very important to provide what is called the "minimum reproducible case". That greatly help to reduce the time to investigate and fix issues. It is also good to develop incrementally, add one dependency at a time, make sure it works, test everything and then move to the next one, instead of adding all at the same time.

The first thing I would try, is that in your default profile the libcxx is ``compiler.libcxx=libstdc++``. Check the output you sent. It is very likely that your default libcxx is actually libstdc++11. This is highlighted as an important item in the "getting started" page of our docs: https://docs.conan.io/en/latest/getting_started.html. I would strongly recommend carefully reading the docs, as they provide a lot of insights that would help. Please try that, and let us know.

**Comment by muhflare** [comment]
you are right, I updated the `compiler.libcxx=libstdc++11` and then ran it again.
but it seems I received the same error of linking, attaching the whole output file
[out.txt](https://github.com/conan-io/conan/files/9009009/out.txt)

**Comment by jcar87** [comment]
Hi @muhflare  - 
could you try adding `-lm` when invoking g++ to compile and link the executable?

```
g++ src/main.cpp @build/conanbuildinfo.args -lm -o bin/argss -std=c++17
                                             ^ here
```

If this works, it might indicate some library in the dependency chain needs to express a dependency on the glibc math library.

**Comment by muhflare** [comment]
I think the error was shortened a little bit, but still failed
[out.txt](https://github.com/conan-io/conan/files/9009790/out.txt)

**Comment by memsharded** [comment]
Based on https://stackoverflow.com/questions/62334452/fast-math-cause-undefined-reference-to-pow-finite, I think it might be a libc incompatibility, that could arise because of that specific LinuxMint distro and fast-math availability

These are the things I would try next:
- Try building from source with ``--build`` (important to remove the other, unused dependencies, to reduce the buid time)
- Try running the same in some docker image, like the conanio/xxxx images, just to rule out if it is a libc issue, or it could be a Conan recipe or tool issue

**Comment by muhflare** [comment]
Great I will give it a shot

Thanks!!!

**Comment by jbcbt** [comment]
FYI, 

I faced the same issue. 
I was also new to conan package manager and was using opencv to investigate its capabilities.
I had attempted to install binaries (ubuntu 20.04)
	`conan install . --install-folder build/`

which resulted in dependency failures (apologies I don't have the error messages), so I 
	`sudo apt-get install`

the relevant dependencies and then repeated the conan install process, until when I believed it was complete.

After this when I attempted to build my project I got the same linker errors i.e. `__log2f_finite` related to the maths library.

I read you last comment in regard to 
	_try building from source with --build (important to remove the other, unused dependencies, to reduce the buid time)_	

and decided to 
	`conan install . --install-folder build/`

again, it spent some time building, after which my project linked. 

For me the issue was _"human error"_.

Thanks for pointing me in the right direction.

**Comment by memsharded** [comment]
Closing this ticket as outdated/staled, never got the feedback about the possible libc incompatibility.
Please create new tickets for any further question, issue or comment. Many thanks for your feedback!
